### PR TITLE
Enhancement: Consider baseURI of referringElement (<base href="...">)

### PR DESCRIPTION
**Pull request by eHtmlu**
#108

### CODE REVIEW

File: lib/Model/ReferringElement.php [hunk 1]
```diff
@@ -8,6 +8,7 @@ class ReferringElement
 {
     private $xpath = '';
     private $title = '';
+    private $baseURI = '';
```
Comment (dantleech) [inline lib/Model/ReferringElement.php:11]: Camel case would be more consistent `$baseUri`

File: lib/Model/ReferringElement.php [hunk 2]
```diff
@@ -19,10 +20,16 @@ public static function fromDOMNode(DOMNode $element): ReferringElement
         $new = new self();
         $new->xpath = $element->getNodePath();
         $new->title = trim($element->nodeValue);
+        $new->baseURI = $element->baseURI;
 
         return $new;
     }
 
+    public function baseURI(): ?string
```
Comment (dantleech) [inline lib/Model/ReferringElement.php:28]: same

File: lib/Model/Url.php [hunk 3]
```diff
@@ -76,7 +76,13 @@ public function resolveUrl(string $link, ReferringElement $referringElement = nu
             throw new InvalidUrl($e->getMessage(), 0, $e);
         }
 
-        $resolvedLink = FollowRedirects::resolve($this->uri, $parsedLink);
+        if ($referringElement && ($baseURI = $referringElement->baseURI())) {
+            $base = self::fromUrl($baseURI);
+        } else {
+            $base = $this;
+        }
```
Comment (dantleech) [inline lib/Model/Url.php:83]: maybe extract to a private method `resolveBaseUri` and remove the `else`?

**Comment by eHtmlu** [comment]
Thank you for all your hints! I tried to make all of these things. But I couldn't get the unit tests to run successfully on my machine. I think, however, my added test should work well. (don't know whether it's already automatically checked here).

**Comment by dantleech** [comment]
Nice work, thanks! The test works fine on Travis and on my local machine, strange that it doesn't on yours :thinking:

**Comment by eHtmlu** [comment]
> Nice work, thanks! The test works fine on Travis and on my local machine, strange that it doesn't on yours 🤔

Maybe I haven't installed it correctly.

If I do this:
```
git clone https://github.com/dantleech/fink.git .
composer install
./vendor/bin/phpunit
```
then I get many errors. It says the command "bin/fink" doesn't exist.

If I call "bin/fink" directly:
```
bin/fink http://example.com/
```
it works well 🤷‍♂️

Anyway, thanks for the merge! I'm happy I could help! 😊 ... was my first time making a pull request 😊

**Pull request merged**